specs/core/1_new_shards.md

```diff
@@ -0,0 +1,440 @@
+            prev_gasprice = state.shard_states[shard].gasprice
+            for i in range(len(offset_slots)):
+                shard_state, block_length, chunks = transition.shard_states[i], transition.shard_block_lengths[i], transition.shard_data_roots[i]
+                block_length = transition.shard
+                assert shard_state.gasprice == update_gasprice(prev_gasprice, block_length)
+                assert shard_state.slot == offset_slots[i]
+                assert len(chunks) == block_length // SHARD_BLOCK_CHUNK_SIZE
+                filled_roots = chunks + [EMPTY_CHUNK_ROOT] * (MAX_SHARD_BLOCK_CHUNKS - len(chunks))
+                assert shard_state.latest_block_hash == hash_tree_root(filled_roots)
+                prev_gasprice = shard_state.gasprice
+
+            # Save updated state
+            state.shard_states[shard] = transition.shard_states[-1]
+            state.shard_states[shard].slot = state.slot - 1
+
+            # Save success (for end-of-epoch rewarding)
+            pending_attestation.crosslink_success = True
+
+            # Apply proposer reward and cost
+            estimated_attester_reward = sum([get_base_reward(state, attester) for attester in attesting_indices])
+            increase_balance(state, proposer, estimated_attester_reward // PROPOSER_REWARD_COEFFICIENT)
+            for state, length in zip(transition.shard_states, transition.shard_block_lengths):
+                decrease_balance(state, proposer, state.gasprice * length)
+            
+    # Type 2: delayed attestations
+    else:
+        assert slot_to_epoch(data.slot) in (get_current_epoch(state), get_previous_epoch(state))
+        assert data.shard_transition_hash == Hash()
+        assert len(attestation.custody_bits) == 0
+
+    for index in attesting_indices:
+        online_countdown[index] = ONLINE_PERIOD
+
+
+    if data.target.epoch == get_current_epoch(state):
+        assert data.source == state.current_justified_checkpoint
+        state.current_epoch_attestations.append(pending_attestation)
+    else:
+        assert data.source == state.previous_justified_checkpoint
+        state.previous_epoch_attestations.append(pending_attestation)
+```
+
+### Misc block post-processing
+
+```python
+def misc_block_post_process(state: BeaconState, block: BeaconBlock):
+    # Verify that a `shard_transition` in a block is empty if an attestation was not processed for it
+    for shard in range(MAX_SHARDS):
+        if state.shard_states[shard].slot != state.slot - 1:
+            assert block.shard_transition[shard] == ShardTransition()
+```
+
+### Light client processing
+
+```python
+def verify_light_client_signatures(state: BeaconState, block: BeaconBlock):
+    period_start = get_current_epoch(state) - get_current_epoch(state) % LIGHT_CLIENT_COMMITTEE_PERIOD
+    committee = get_light_client_committee(state, period_start - min(period_start, LIGHT_CLIENT_COMMITTEE_PERIOD))
```

Comment:
Why the `min` here?

Shouldn't it just be:

```suggestion
    committee = get_light_client_committee(state, period_start)
```

Reply:
We need to have logic somewhere to say "the light client committee used at slot N takes data from slot N - N % PERIOD - PERIOD (with one special case where N < PERIOD to avoid underflow). I just fixed; moved all that logic into `get_light_client_committee` itself.